Paper: Semi-Supervised Semantic Annotator (S3A): Toward Efficient Semantic Image Labeling

ntjess opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@cbcunc<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@jameslamb<!--end-reviewers-list-->

ntjess committed:
s3a: initial commit

ntjess committed:
S3A intro + title formatting, raw latex style

ntjess committed:
add builder/overleaf syncup script

ntjess committed:
221_jessurun: first draft complete

ntjess committed:
Better overleaf syncing
- Sync references.bib & main.tex along with sections and figures
- Delete about-to-sync directories to guarantee proper checkouts

scoobies commented:
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

ntjess commented:
Marked as draft since video link is still not created

ntjess committed:
Reformat using RST files

deniederhut commented:
The video link won't exist until after the conference -- feel free to remove the draft designation if you like

ntjess commented:
Thank you for the clarification. If my paper is a virtual poster session, am I able to record my own (non-SciPy) video which walks through the material & relevant Python demos? Or is the slot dedicated to oral presentation links?

Apologies if the answer is readily available online and I missed it.

ntjess committed:
Remove unneeded video link

ntjess committed:
More readable tex2rst regex patterns, search for "\texttt{...}"

ntjess committed:
remove outdated 'case_study.rst

deniederhut commented:
> If my paper is a virtual poster session, am I able to record my own (non-SciPy) video which walks through the material & relevant Python demos?

I'm afraid I have no idea. Would you mind reaching out to the program committee to ask?

deniederhut commented:
@scoobies mark paper

deniederhut commented:
@scoobies build status

scoobies commented:
{"ntjess-s3a_paper"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmpkwvbgdsc'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: 221_jessurun\n*** Warning: /tmp/tmpkwvbgdsc/publisher/../output/221_jessurun/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/ntjess-s3a_paper.pdf", "build_status"=>"success", "build_timestamp"=>"27/05 15:55"}, "status"=>"success"}}

deniederhut commented:
@scoobies mark needs review

deniederhut commented:
@scoobies assign @cbcunc as editor

scoobies commented:
Assigned! @cbcunc is now the editor

ntjess committed:
221_jessurun: Add poster draft

cbcunc commented:
Editor confirms paper builds.

deniederhut commented:
@scoobies add @cx0 to reviewers

scoobies commented:
@cx0 added to the reviewers list!

deniederhut commented:
Hi! We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

deniederhut commented:
@scoobies add @jameslamb to reviewers

scoobies commented:
@jameslamb added to the reviewers list!

cbcunc commented:
@cx0, @jameslamb First review is due June 16.

jameslamb commented on papers/221_jessurun/old rst markup:
```diff
@@ -0,0 +1,57 @@
+:author: Gaius Caesar
```
Given the file name and what appears to be filler text from a template, it seems to me that this file was left behind from development. Can it be removed?

ntjess replied:
Indeed, this is an old template and was deleted

jameslamb commented on papers/221_jessurun/sections_rst/intro.rst:
```diff
@@ -0,0 +1,79 @@
+
+
+..
+    %One such field applying this method include Bill-of-Material (BoM) extraction. This BoM, or list of all surface-mount devices (SMDs) on a printed circuit board (PCB) surface, can be generated from optical images of a board under test. Next, it can be compared against a reference design to detect likely counterfeit, tampered, or defective SMDs :cite:`paradis2020color,azhaganReviewAutomaticBil ...
```
```suggestion
    %One such field applying this method includes Bill-of-Material (BoM) extraction. This BoM, or list of all surface-mount devices (SMDs) on a printed circuit board (PCB) surface, can be generated from optical images of a board under test. Next, it can be compared against a reference design to detect likely counterfeit, tampered, or defective SMDs :cite:`paradis2020color,azhaganReviewAutomaticBill2019`. To create such a BoM using optical data alone, it is crucial that detected SMDs use segmentation masks that are representative of structures such as pins, solder flow, and more. This is the only way the collected data will remain indicative of true BoM properties. An example annotation is shown in Figure :ref:`figpcb`.
```

I've suggested one typo change here. It would be easier to see if these `.rst` files were reformatted following the suggestions in https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md#other-requirements

> Avoid long paragraphs of your rich text file to enable reviewers to target specific sections with their comments. GitHub comments are per line of text.

jameslamb replied:
^ and I should have included in my review summary...this suggestion to break multi-sentence paragraphs into individual lines, for purpose of improving the review process, extends to every file in this pull request

ntjess replied:
Paragraphs split with newlines per recommendation

jameslamb commented on papers/221_jessurun/sections_rst/intro.rst:
```diff
@@ -0,0 +1,79 @@
+Introduction
+============
+Labeled image data is essential for tuning and evaluating the performance of machine learning applications.
```
```suggestion
Labeled image data is essential for training, tuning, and evaluating the performance of many machine learning applications.
```

I recommend adding "training" to the list of tasks, and changing "machine learning applications" to "many machine learning applications". As written, this sentence makes it sound like this paper is relevant to every machine learning project, which is not true...a wide range of machine learning tasks do not rely on image data at all, such as regression/ranking/classification using tabular data.

jameslamb commented on papers/221_jessurun/sections_rst/intro.rst:
```diff
@@ -0,0 +1,79 @@
+Introduction
+============
+Labeled image data is essential for tuning and evaluating the performance of machine learning applications.
+Such labels are typically defined with approximate enclosing shapes (i.e. simple polygons or parametric shapes), which tend to misrepresent more complex components.
```
```suggestion
Such labels are typically defined with approximate enclosing shapes (e.g. "this square contains a cat" ), which tend to misrepresent more complex components (e.g. "the cat only occupies 70% of the square").
```

I suspect that a wide, interdisciplinary audience might not know what "parametric shapes" means. Reading this paper is the first time I'd ever heard that phrase.

Per https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md#written-quality

> highly domain-specific jargon should be avoided or explained where possible

Would you consider changing this sentence to something like what I've suggested above, with plain-language descriptions of the topic? I think it might help a wider audience to understand the importance of what is being proposed.

jameslamb commented on papers/221_jessurun/sections_rst/intro.rst:
```diff
@@ -0,0 +1,79 @@
+
+
+..
+    % Especially when each region can be arbitrarily complex, the software will enter a non-responsive state where no annotation can be performed.
```
```suggestion
    % Especially when each region can be arbitrarily complex, the software can enter a non-responsive state where no annotation can be performed.
```

jameslamb commented on papers/221_jessurun/sections_rst/methods.rst:
```diff
@@ -0,0 +1,59 @@
+
+Application Overview
+====================
+Design decisions throughout S3A's architecture have been driven by the following objectives: metadata should have significance rather than be treated as an afterthought; high-resolution images should have minimal impact on the annotation workflow; ROI density and complexity should not limit annotation workflow; and prototyping should not be hindered by application complexity.
```
```suggestion
Design decisions throughout S3A's architecture have been driven by the following objectives:

* metadata should have significance rather than be treated as an afterthought
* high-resolution images should have minimal impact on the annotation workflow
* ROI density and complexity should not limit annotation workflow
* prototyping should not be hindered by application complexity
```

Would you consider making this a bulleted list? I think that will be easier for readers to absorb than the current long sentence with items separated by semicolons.

jameslamb commented on papers/221_jessurun/sections_rst/methods.rst:
```diff
@@ -0,0 +1,59 @@
+
+Plugins for User Extensions
+---------------------------
+The previous section briefly described how custom user functions are easily be wrapped within a process, exposing its parameters within S3A in a GUI format. A rich plugin interface is built on top of this capability in which custom functions, table field predictors, default action hooks, and more can be directly integrated into S3A. In all cases, only a few lines of code are required to achieve m ...
```
```suggestion
The previous section briefly described how custom user functions are easily wrapped within a process, exposing its parameters within S3A in a GUI format. A rich plugin interface is built on top of this capability in which custom functions, table field predictors, default action hooks, and more can be directly integrated into S3A. In all cases, only a few lines of code are required to achieve most integrations between user code and plugin interface specifications. The core plugin infrastructure consists of a function/property registration mechanism and an interaction window that shows them in the UI. As such, arbitrary user functions can be `registered` in one line of code to a plugin, where it will be effectively exposed to the user within S3A.
```

jameslamb commented on papers/221_jessurun/sections_rst/methods.rst:
```diff
@@ -0,0 +1,59 @@
+
+Adaptable I/O
+-------------
+An extendable I/O framework allows annotations to be used in a myriad of ways. Out-of-the-box, S3A easily supports instance-level segmentation outputs, facilitating deep learning model training. As an example, Figure :ref:`figcropExports` illustrates how each instance in the image becomes its own pair of image and mask data. When several instances overlap, each is uniquely distinguishable dependi ...
```
Would you consider adding links to COCO and Pascal VOC? I had never heard of those file formats before, and don't believe they're as ubiquitous as, say, JSON or CSV.

I don't know whether or not a citation is required for them...but at least, as a reader who might be new to these technologies, I'd appreciate links to documentation explaining what those things are.

jameslamb commented on papers/221_jessurun/sections_rst/casestudy.rst:
```diff
@@ -0,0 +1,70 @@
+
+
+..
+    % PCB example: In PCB assurance, it can be desirable to semantically annotate PCB components $\rightarrow$ e.g. component footprint analysis. Some compoenents have complex shapes and even holes $\rightarrow$ Complex Vertices case. For many other annotation tools, complex annotations either cannot be done (annotator only supports bounding boxes), must be done completely manually (point-by-poin ...
```
Without the PCB-specific language, this section is a VERY powerful, plain-language description of why this software exists and what problems it solves. Would you consider putting some of these specifics in the introduction as well?

The examples like "only supports bounding boxes", "doesn't support holes", etc. could be understand clearly by a wide audience.

jameslamb commented on papers/221_jessurun/sections_rst/casestudy.rst:
```diff
@@ -0,0 +1,70 @@
+
+
+..
+    % show off screenshot of complex annotation(s) such as a 64-pin IC, large jack, or large connector - bonus if there's an annotation with holes in it bc idk many other annotation apps that allow annotations to have holes -- TODO: insert fig of transistor with holes in heat sink 
```
Noting for you that this `TODO` has not yet been addressed, in case this was forgotten.

jameslamb commented on papers/221_jessurun/sections_rst/casestudy.rst:
```diff
@@ -0,0 +1,70 @@
+
+
+..
+    % Extension Example: Though this aspect of S3A was inspired by PCB component footprint analysis, there are certainly other applications that deal with complex vertices such as in biomedical, Y, and Z. -- Might not need to be justified since it's a case study, but I'm definitely open to alternative suggestions
```
Noting that placeholders "Y" and "Z" have not yet been filled in, and that there is still this "Might not need to be justified" comment.

jameslamb commented on papers/221_jessurun/sections_rst/casestudy.rst:
```diff
@@ -0,0 +1,70 @@
+
+
+..
+    % PCB example: In PCB assurance, it can be desirable to associate metadata with elements on a PCB $\rightarrow$ e.g. design rule checking. For example, annotating reference designators, it can be useful to note the text values, text orientation, and any components the reference designator is referrring to $\rightarrow$ complex metadata case. For many other annotation tools, complex metadata e ...
```
Noting that this still has placeholders "X, Y, and Z" in it.

jameslamb commented on papers/221_jessurun/sections_rst/casestudy.rst:
```diff
@@ -0,0 +1,70 @@
+
+
+..
+    % show off screenshot of PCB OCR metadata and/or yaml file/table of different types of metadata you can collect. especially highlight the comp locator for complex reference designators (e.g. designators that refer to multiple comps such as R1-R5) -- TODO: insrt fig of yaml file for OCR and link to https://pyqtgraph.readthedocs.io/en/latest/parametertree/parametertypes.html for extensions 
```
Noting that this looks like a "note to self" type of TODO

jameslamb commented on papers/221_jessurun/sections_rst/casestudy.rst:
```diff
@@ -0,0 +1,70 @@
+
+
+..
+    % Extension Example: Though this aspect of S3A was inspired by PCB design rule checking, there are certainly other applications that deal with complex vertices such as in human annotation, Y, and Z. 
```
Noting that this still has placeholders "Y" and "Z" in it.

jameslamb commented on papers/221_jessurun/references.bib:
```diff
@@ -0,0 +1,573 @@
+  issn = {0957-4174},
+  doi = {10.1016/j.eswa.2014.12.045},
+  abstract = {Multilayered aseptic material which guarantees the extended shelf life of liquid foods is turned into an impermeable package by folding packaging machines. Some problems in the machines causes the packages to lose the impermeability property and therefore lead to deterioration of the liquid food. The leak test, which was performed to determine the problem, has been carried out by ex ...
+  file = {/home/ntjess/Zotero/storage/S4B347RU/Adem et al. - 2015 - Image processing based quality control of the impe.pdf;/home/ntjess/Zotero/storage/LE9JH7HZ/S0957417414008240.html},
```
It's been a long time since I've used LaTeX, but I generally don't expect to see fully-specific absolute file paths in any file checked into source control for general consumption. Is this filepath specific to one of the author's development environments, or to the conference's build system? Will having it here cause issues for others wanting to render these documents?

jameslamb commented on papers/221_jessurun/figures/makefigs.tex:
```diff
@@ -0,0 +1,107 @@
+\end{figure}
+}
+
+% \newcommand{\makeSegComparisonFig}{%
```
Noting this commented-out code, which should either be removed or uncommented. Commented-out code leads to false positives for code search and linters, and increased cognitive load for those editing these files.

jameslamb commented on papers/221_jessurun/sections/methods.tex:
```diff
@@ -0,0 +1,31 @@
+\subsection{Plugins for User Extensions}\label{sec:plugins}
+The previous section briefly described how custom user functions are easily be wrapped within a process, exposing its parameters within S3A in a GUI format. A rich plugin interface is built on top of this capability in which custom functions, table field predictors, default action hooks, and more can be directly integrated into S3A. In all cases, only a few lines of code are required to achieve m ...
+
+% \makeCustomMiscFuncFig
```
It seems that `makeCustomMiscFuncFig` hasn't actually been included in this PR. Noting this to remind the authors to please include a figure here.

Since this paper is being submitted to a Python conference, I also strongly recommend adding a code snippet that can be copied and pasted by users of S3A. This document claims that customizing S3A is as easy as "a few lines of code".

Including a few lines of code that provide some useful customization would be a powerful way to prove that to skeptical readers.

ntjess replied:
Thank you for your recommendation. I added a new figure and code snippet (Figure 5 -- `figures/funcIntegration.pdf`), is this a sufficient representation? I'm happy to adjust these examples if it would add more clarity.

jameslamb replied:
Thanks very much for that! Yes, I think the snippet you've linked to (https://gitlab.com/s3a/s3a/-/snippets/2203628) sufficiently addresses this review suggestion.

jameslamb reviewed (changes requested):
Thanks very much for the opportunity to review this paper, and for all of the hard work you put into it!

I left some minor comments, but don't have any suggested large structural changes.

### Notes for Scipy Proceedings Committee

I recommend inclusion of this paper in the proceedings.

You may want to weight my recommendation less than recommendations of other reviewers, as I have very little experience with computer vision and therefore with image labelling. Some of the file formats mentioned by the authors were new to me, and this paper is the first time I've ever heard the phrase "semantic segmentation".

However, I do feel that I understand the problem that S3A solves and how it solves it. I have faced similar problems in the past with geospatial point and shapefile data....holes, highly-complex boundaries, etc. and in those situations usually just did some of the suboptimal things described as the current state in this paper, like using polygon simplification or just accepting some point-level inaccuracies in exchange for drawing simple bounding polygons like boxes or hexagons. I've also written Qt plugins before for Orange (a piece of software cited by this paper), and can appreciate the importance of the ecosystem of project around Qt to scientific computing in Python.

Thanks very much for your time and consideration.

ntjess commented:
@jameslamb I appreciate the thorough review and valuable feedback! My apologies for the administrative issues (i.e. commented-out code, single-line paragraphs); this is the first conference I've submitted to where source documents were reviewed instead of just the output PDF.

deniederhut commented:
Thanks James!

deniederhut commented:
@scoobies mark needs more review

ntjess committed:
Clean up administrative issues:
- Commented-out code is removed
- Paragraphs are broken by line for easier review
- Old template file deleted
- "file" hardlinks in references

deniederhut commented:
@scoobies remove @cx0 from reviewers

scoobies commented:
@cx0 removed from the reviewers list!

deniederhut commented:
@scoobies mark pending comment

ntjess committed:
Address jameslamb's comments

Commit:
avoid duplicate usage of `subsequently`

deniederhut commented:
Thanks for the revisions @ntjess ! Is this paper ready for re-review?

ntjess committed:
Minor update to author name

ntjess commented:
@deniederhut no problem! Yes -- the draft is ready for additional review. I wasn't sure if I needed to post anything after revisions were tentatively completed.

deniederhut commented:
Awesome! @jameslamb will you have time to take a second look at the paper this week?

cbcunc commented:
Reviewers, please give your final recommendations. The deadline for final review comments is tomorrow.

Authors, please respond to review comments with changes or comments which resolve the conversation. Final acceptance decisions are July 4.

jameslamb commented:
>  @jameslamb will you have time to take a second look at the paper this week?
> ...
> the deadline for final review comments is tomorrow

I can look again tonight (Chicago time, so 10-12 hours from now).

jameslamb reviewed (approved):
I reviewed this again tonight (just by looking at commits made since my initial review) and do not have any additional comments. Very interesting project!

cbcunc commented:
@scoobies mark ready
